src/app/binding/BindingTable.h

```diff
@@ -0,0 +1,104 @@
+#include "lib/dnssd/Resolver.h"
+#include "lib/support/Pool.h"
+
+namespace chip {
+namespace app {
+
+struct BindingTableEntry
+{
+    BindingTableEntry(EndpointId sourceEndpoint, NodeId peerNodeId, EndpointId peerEndpointId, FabricIndex peerFabricIndex,
+                      ClusterId clusterId);
+
+    EndpointId mSourceEndpoint;
+    EndpointId mPeerEndpointId;
+    PeerId mPeerId;
+    GroupId mPeerGroupId; // Destination group, not used
+    ClusterId mClusterId;
+};
+
+class BindingTable
+{
+public:
+    BindingTable();
+
+    CHIP_ERROR Add(EndpointId sourceEndpoint, NodeId peerNodeId, EndpointId peerEndpointId, FabricIndex peerFabricIndex,
+                   ClusterId clusterId);
+
+    CHIP_ERROR Remove(EndpointId sourceEndpoint, NodeId peerNodeId, EndpointId peerEndpointId, FabricIndex peerFabricIndex,
+                      ClusterId clusterId);
+
+    BindingTableEntry * FindBindingEntry(PeerId peer);
+
+    BindingTableEntry * FindBindingEntry(NodeId node, FabricIndex fabricIndex);
+
+    BindingTableEntry * FindBindingEntry(EndpointId endpoint, ClusterId cluster);
+
+    OperationalDeviceProxy * FindPeerDevice(PeerId peer);
+
+    OperationalDeviceProxy * FindPeerDevice(NodeId node, FabricIndex fabricIndex);
+
+    CHIP_ERROR TryConnectDevice(OperationalDeviceProxy * device);
+
+    static BindingTable & GetInstance() { return BindingTable::sInstance; }
+
+private:
+    static BindingTable sInstance;
+    static constexpr int kMaxBinding = 3; // TODO: make it (3 * NUM_SUPPORTED_FABRICS)
+
+    class ResolveDelegate : public Dnssd::ResolverDelegate
+    {
+    public:
+        ResolveDelegate(BindingTable * bindingTable) : mBindingTable(bindingTable) {}
+
+        void OnNodeIdResolved(const Dnssd::ResolvedNodeData & nodeData) override;
+
+        void OnNodeIdResolutionFailed(const PeerId & peerId, CHIP_ERROR error) override;
+
+        void OnNodeDiscoveryComplete(const Dnssd::DiscoveredNodeData & nodeData) override {}
+
+    private:
+        BindingTable * mBindingTable;
+    };
+
+    static void OnDeviceConnectedFn(void * context, chip::DeviceProxy * device);
+    static void OnDeviceConnectionFailureFn(void * context, NodeId deviceId, CHIP_ERROR error);
+
+    BitMapObjectPool<BindingTableEntry, kMaxBinding> mBindingEntries;
+    BitMapObjectPool<OperationalDeviceProxy, kMaxBinding> mPeerDevices;
+    ResolveDelegate mResolveDelegate;
+    Callback::Callback<OnDeviceConnected> mOnDeviceConnected;
+    Callback::Callback<OnDeviceConnectionFailure> mOnDeviceConnectionFailure;
+};
+
+void MatterBindingAttributeChangeCallback(const chip::app::ConcreteAttributePath & attributePath, uint8_t mask, uint8_t type,
+                                          uint16_t size, uint8_t * value);
+
+} // namespace app
+} // namespace chip
+
+// User callback called when a bound cluster is changed. The peer device will be already connected.
+extern void MatterBoundAttributeChangedCallback(const chip::app::BindingTableEntry & binding,
```

Comment:
Instead of a free floating function symbol like this, can you please instead define a delegate/callback interface on the Binding cluster implementation that can be used to notify the application of changes?